src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -27,178 +25,78 @@ public class CompilerCache : ICompilerCache
         /// Initializes a new instance of <see cref="CompilerCache"/> populated with precompiled views
         /// discovered using <paramref name="provider"/>.
         /// </summary>
-        /// <param name="assemblyProvider">The <see cref="IAssemblyProvider"/> that provides assemblies
-        /// for precompiled view discovery.</param>
+        /// <param name="razorFileInfoCollections">The sequence of <see cref="RazorFileInfoCollection"/> that provides
+        /// information for precompiled view discovery.</param>
         /// <param name="loaderContextAccessor">The <see cref="IAssemblyLoadContextAccessor"/>.</param>
         /// <param name="optionsAccessor">An accessor to the <see cref="RazorViewEngineOptions"/>.</param>
-        public CompilerCache(IAssemblyProvider assemblyProvider,
-                             IAssemblyLoadContextAccessor loadContextAccessor,
-                             IOptions<RazorViewEngineOptions> optionsAccessor)
-            : this(GetFileInfos(assemblyProvider.CandidateAssemblies),
-                  loadContextAccessor.GetLoadContext(RazorFileInfoCollectionType.Assembly),
+        public CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
+            IAssemblyLoadContextAccessor loadContextAccessor,
+            IOptions<RazorViewEngineOptions> optionsAccessor)
+            : this(razorFileInfoCollections,
+                  loadContextAccessor.GetLoadContext(typeof(CompilerCache).GetTypeInfo().Assembly),
                   optionsAccessor.Options.FileProvider)
         {
         }
 
-        internal CompilerCache(IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
-                               IAssemblyLoadContext loadContext,
-                               IFileProvider fileProvider)
+        internal CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
+            IAssemblyLoadContext loadContext,
+            IFileProvider fileProvider)
```

Comment:
Does all of this work need to happen on construction? This causes cascading... services don't want to be initialized because they trigger loading of all the views. 

What we do for things like ADs is expose a 'Collection' (not a service) via a 'Collection Holder' which is a service. Then anyone who wants to can DI the service without forcing eager evaluation. Accessing the collection will trigger all the init code.